### PR TITLE
Add products command

### LINKED ISSUE #466: a command to display available, staged, and deployed products alongside each other
as a platform engineer, i'd like a command that gives me a combined view of available, staged, and deployed products.

something along these lines:

om_products

i just forked this repo today and started working on an implementation in a branch named `products-command`.

would love your feedback.  let me know if that's something that'd be worth incorporating as an additional om subcommand.

### PR DESCRIPTION
**Pull request by iplay88keys**
* Lists available, staged, and deployed products
* Replaces 'available-products', 'staged-products', and 'deployed-products' commands
* Update docs around products command
* Add deprecation notices to 'available-products', 'staged-products', and 'deployed-products' commands

Resolves #466

### CODE REVIEW

**Comment by iplay88keys** [comment]
Example of base products command as well as one of the options for setting the columns (--deployed). Supports any combination of `--available`, `--staged`, and `--deployed` with the default of all columns. I also sorted the output of the product names.

image

**Comment by iplay88keys** [comment]
There are probably some edge cases that need to be thought through.

**Comment by jtarchie** [comment]
We just tried with `available-products` on our test environment.

With `-a`,

```
+---------------------------+----------------+
|           NAME            |   AVAILABLE    |
+---------------------------+----------------+
| cf                        | 2.9.8          |
| p-bosh                    |                |
| p-healthwatch             | 1.8.2-build.1  |
| pivotal-container-service | 1.7.0-build.26 |
| pivotal-telemetry-om      | 1.1.1-build.5  |
+---------------------------+----------------+
```

When we run it with the command `available-products`,

```
+---------------------------+----------------+
|           NAME            |    VERSION     |
+---------------------------+----------------+
| p-healthwatch             | 1.8.1-build.1  |
| pivotal-telemetry-om      | 1.1.0-build.4  |
| p-healthwatch             | 1.8.2-build.1  |
| pivotal-container-service | 1.7.0-build.26 |
| cf                        | 2.9.8          |
| pivotal-telemetry-om      | 1.1.1-build.5  |
+---------------------------+----------------+
```

There is a difference of:
* the `-a` shows `p-bosh`, the other does not
* the command `available-products` shows to versions of `p-healthwatch`

**Comment by iplay88keys** [comment]
Cool, I'll look into those and push a change with the fixes!

**Comment by kcboyle** [comment]
Keep in mind the combined view @eitansuez linked in the issue for as well (the empty line so that everything organizes nice)
so for available products:
```
+---------------------------+----------------+
|           NAME            |    VERSION     |
+---------------------------+----------------+
| p-healthwatch             | 1.8.1-build.1  |
|                           | 1.8.2-build.1  |
| pivotal-telemetry-om      | 1.1.0-build.4  |
| pivotal-container-service | 1.7.0-build.26 |
| cf                        | 2.9.8          |
| pivotal-telemetry-om      | 1.1.1-build.5  |
+---------------------------+----------------+
```

**Comment by iplay88keys** [comment]
Here's the updated output for `om products`:

**Comment by iplay88keys** [comment]
Here's the updated output for `om products`:
```
+-----------------------------+-----------------+-----------------+-----------------+
|            NAME             |    AVAILABLE    |     STAGED      |    DEPLOYED     |
+-----------------------------+-----------------+-----------------+-----------------+
| appMetrics                  | 2.0.6-dev.005   | 2.0.6-dev.005   | 2.0.6-dev.005   |
| cf                          | 2.8.16          | 2.8.16          | 2.8.16          |
| metric-store                | 1.4.4           | 1.4.4           | 1.4.4           |
| p-event-alerts              | 1.2.9-build.1   | 1.2.9-build.1   | 1.2.9-build.1   |
| p-healthwatch               | 1.8.2-build.1   | 1.8.3-build.3   | 1.8.2-build.1   |
|                             | 1.8.3-build.3   |                 |                 |
| p-healthwatch2              | 2.0.5-build.84  | 2.0.5-build.84  | 2.0.5-build.84  |
| p-healthwatch2-pas-exporter | 2.0.5-build.84  | 2.0.5-build.84  | 2.0.5-build.84  |
| p-isolation-segment         | 2.8.2           | 2.8.2           | 2.8.2           |
| p-rabbitmq                  | 1.19.1-build.22 | 1.19.1-build.22 | 1.19.1-build.22 |
| p-redis                     | 2.3.1-build.36  | 2.3.1-build.36  | 2.3.1-build.36  |
| pivotal-mysql               | 2.8.0-build.111 | 2.8.0-build.111 | 2.8.0-build.111 |
| pivotal-telemetry-om        | 1.0.1-build.2   | 1.0.1-build.2   | 1.0.1-build.2   |
+-----------------------------+-----------------+-----------------+-----------------+
```

And `om products -a`:
```
+-----------------------------+-----------------+
|            NAME             |    AVAILABLE    |
+-----------------------------+-----------------+
| appMetrics                  | 2.0.6-dev.005   |
| cf                          | 2.8.16          |
| metric-store                | 1.4.4           |
| p-event-alerts              | 1.2.9-build.1   |
| p-healthwatch               | 1.8.2-build.1   |
|                             | 1.8.3-build.3   |
| p-healthwatch2              | 2.0.5-build.84  |
| p-healthwatch2-pas-exporter | 2.0.5-build.84  |
| p-isolation-segment         | 2.8.2           |
| p-rabbitmq                  | 1.19.1-build.22 |
| p-redis                     | 2.3.1-build.36  |
| pivotal-mysql               | 2.8.0-build.111 |
| pivotal-telemetry-om        | 1.0.1-build.2   |
+-----------------------------+-----------------+
```

**Comment by iplay88keys** [comment]
Seems like I dropped the `p-bosh` from the overall product view. I'll look into why that is.

**Comment by kcboyle** [comment]
All Comparison outputs (for documentation purposes):
```
→ go run main.go --env ~/workspace/platform-automation-deployments/reference-gcp/env.yml deployed-products
+---------------------------+-----------------+
|           NAME            |     VERSION     |
+---------------------------+-----------------+
| pivotal-container-service | 1.7.0-build.26  |
| cf                        | 2.9.8           |
| p-healthwatch             | 1.8.1-build.1   |
| p-bosh                    | 2.9.6-build.148 |
| pivotal-telemetry-om      | 1.1.1-build.5   |
+---------------------------+-----------------+

→ go run main.go --env ~/workspace/platform-automation-deployments/reference-gcp/env.yml products -d
+---------------------------+-----------------+
|           NAME            |    DEPLOYED     |
+---------------------------+-----------------+
| cf                        | 2.9.8           |
| p-bosh                    | 2.9.6-build.148 |
| p-healthwatch             | 1.8.1-build.1   |
| pivotal-container-service | 1.7.0-build.26  |
| pivotal-telemetry-om      | 1.1.1-build.5   |
+---------------------------+-----------------+

→ go run main.go --env ~/workspace/platform-automation-deployments/reference-gcp/env.yml available-products
+---------------------------+----------------+
|           NAME            |    VERSION     |
+---------------------------+----------------+
| p-healthwatch             | 1.8.1-build.1  |
| pivotal-telemetry-om      | 1.1.0-build.4  |
| p-healthwatch             | 1.8.2-build.1  |
| pivotal-container-service | 1.7.0-build.26 |
| cf                        | 2.9.8          |
| pivotal-telemetry-om      | 1.1.1-build.5  |
+---------------------------+----------------+

→ go run main.go --env ~/workspace/platform-automation-deployments/reference-gcp/env.yml products -a
+---------------------------+----------------+
|           NAME            |   AVAILABLE    |
+---------------------------+----------------+
| cf                        | 2.9.8          |
| p-healthwatch             | 1.8.1-build.1  |
|                           | 1.8.2-build.1  |
| pivotal-container-service | 1.7.0-build.26 |
| pivotal-telemetry-om      | 1.1.0-build.4  |
|                           | 1.1.1-build.5  |
+---------------------------+----------------+

→ go run main.go --env ~/workspace/platform-automation-deployments/reference-gcp/env.yml staged-products
+---------------------------+-----------------+
|           NAME            |     VERSION     |
+---------------------------+-----------------+
| pivotal-container-service | 1.7.0-build.26  |
| cf                        | 2.9.8           |
| p-healthwatch             | 1.8.1-build.1   |
| p-bosh                    | 2.9.6-build.148 |
| pivotal-telemetry-om      | 1.1.1-build.5   |
+---------------------------+-----------------+

→ go run main.go --env ~/workspace/platform-automation-deployments/reference-gcp/env.yml products -s
+---------------------------+-----------------+
|           NAME            |     STAGED      |
+---------------------------+-----------------+
| cf                        | 2.9.8           |
| p-bosh                    | 2.9.6-build.148 |
| p-healthwatch             | 1.8.1-build.1   |
| pivotal-container-service | 1.7.0-build.26  |
| pivotal-telemetry-om      | 1.1.1-build.5   |
+---------------------------+-----------------+

 → go run main.go --env ~/workspace/platform-automation-deployments/reference-gcp/env.yml products
+---------------------------+----------------+-----------------+-----------------+
|           NAME            |   AVAILABLE    |     STAGED      |    DEPLOYED     |
+---------------------------+----------------+-----------------+-----------------+
| cf                        | 2.9.8          | 2.9.8           | 2.9.8           |
| p-bosh                    |                | 2.9.6-build.148 | 2.9.6-build.148 |
| p-healthwatch             | 1.8.1-build.1  | 1.8.1-build.1   | 1.8.1-build.1   |
|                           | 1.8.2-build.1  |                 |                 |
| pivotal-container-service | 1.7.0-build.26 | 1.7.0-build.26  | 1.7.0-build.26  |
| pivotal-telemetry-om      | 1.1.0-build.4  | 1.1.1-build.5   | 1.1.1-build.5   |
|                           | 1.1.1-build.5  |                 |                 |
+---------------------------+----------------+-----------------+-----------------+
```

It looks real good (in terms of output) IMO. It takes the same display and functionality, and makes it easier to read and look at. Thanks for alphabetizing the products, too. that adds a lot.

File: commands/products.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+				Name: product.Name,
+			}
+		}
+
+		productVersions := productVersionsCombiner[product.Name]
+		productVersions.Staged = product.Version
+		productVersionsCombiner[product.Name] = productVersions
+	}
+
+	for _, product := range deployedProducts {
```
Comment (kcboyle) [inline commands/products.go:76]: This whole section is clear, and easy to read. If I'm trying to be nitpicky, I'd love to see these `for` loops in functions (`combineStagedProducts`, `combineAvailableProducts`, `combineDeployedProducts` or something similar)
Just to let this flow top->bottom a little faster without the for loop break

Comment (kcboyle) [reply commands/products.go:76]: if these functions also have the `sp` receiver like `Execute`, you can also clean up some of the earlier setup

**Comment by kcboyle** [comment]
Added a couple more display check tests. Adding the outputs here for documentation
Notes:
- the list headers will display in alphabetical order, no matter what order you give the flags in 
- combining flags is 100% allowed
- duplicated flags don't break
```
→ go run main.go --env ~/workspace/platform-automation-deployments/reference-gcp/env.yml products --staged --deployed
+---------------------------+-----------------+-----------------+
|           NAME            |     STAGED      |    DEPLOYED     |
+---------------------------+-----------------+-----------------+
| cf                        | 2.9.8           | 2.9.8           |
| p-bosh                    | 2.9.6-build.148 | 2.9.6-build.148 |
| p-healthwatch             | 1.8.1-build.1   | 1.8.1-build.1   |
| pivotal-container-service | 1.7.0-build.26  | 1.7.0-build.26  |
| pivotal-telemetry-om      | 1.1.1-build.5   | 1.1.1-build.5   |
+---------------------------+-----------------+-----------------+

→ go run main.go --env ~/workspace/platform-automation-deployments/reference-gcp/env.yml products --staged --available
+---------------------------+----------------+-----------------+
|           NAME            |   AVAILABLE    |     STAGED      |
+---------------------------+----------------+-----------------+
| cf                        | 2.9.8          | 2.9.8           |
| p-bosh                    |                | 2.9.6-build.148 |
| p-healthwatch             | 1.8.1-build.1  | 1.8.1-build.1   |
|                           | 1.8.2-build.1  |                 |
| pivotal-container-service | 1.7.0-build.26 | 1.7.0-build.26  |
| pivotal-telemetry-om      | 1.1.0-build.4  | 1.1.1-build.5   |
|                           | 1.1.1-build.5  |                 |
+---------------------------+----------------+-----------------+

→ go run main.go --env ~/workspace/platform-automation-deployments/reference-gcp/env.yml products --available --deployed
+---------------------------+----------------+-----------------+
|           NAME            |   AVAILABLE    |    DEPLOYED     |
+---------------------------+----------------+-----------------+
| cf                        | 2.9.8          | 2.9.8           |
| p-bosh                    |                | 2.9.6-build.148 |
| p-healthwatch             | 1.8.1-build.1  | 1.8.1-build.1   |
|                           | 1.8.2-build.1  |                 |
| pivotal-container-service | 1.7.0-build.26 | 1.7.0-build.26  |
| pivotal-telemetry-om      | 1.1.0-build.4  | 1.1.1-build.5   |
|                           | 1.1.1-build.5  |                 |
+---------------------------+----------------+-----------------+

→ go run main.go --env ~/workspace/platform-automation-deployments/reference-gcp/env.yml products --deployed --staged
+---------------------------+-----------------+-----------------+
|           NAME            |     STAGED      |    DEPLOYED     |
+---------------------------+-----------------+-----------------+
| cf                        | 2.9.8           | 2.9.8           |
| p-bosh                    | 2.9.6-build.148 | 2.9.6-build.148 |
| p-healthwatch             | 1.8.1-build.1   | 1.8.1-build.1   |
| pivotal-container-service | 1.7.0-build.26  | 1.7.0-build.26  |
| pivotal-telemetry-om      | 1.1.1-build.5   | 1.1.1-build.5   |
+---------------------------+-----------------+-----------------+

→ go run main.go --env ~/workspace/platform-automation-deployments/reference-gcp/env.yml products --deployed --staged --deployed
+---------------------------+-----------------+-----------------+
|           NAME            |     STAGED      |    DEPLOYED     |
+---------------------------+-----------------+-----------------+
| cf                        | 2.9.8           | 2.9.8           |
| p-bosh                    | 2.9.6-build.148 | 2.9.6-build.148 |
| p-healthwatch             | 1.8.1-build.1   | 1.8.1-build.1   |
| pivotal-container-service | 1.7.0-build.26  | 1.7.0-build.26  |
| pivotal-telemetry-om      | 1.1.1-build.5   | 1.1.1-build.5   |
+---------------------------+-----------------+-----------------+

→ go run main.go --env ~/workspace/platform-automation-deployments/reference-gcp/env.yml products --deployed --staged --available
+---------------------------+----------------+-----------------+-----------------+
|           NAME            |   AVAILABLE    |     STAGED      |    DEPLOYED     |
+---------------------------+----------------+-----------------+-----------------+
| cf                        | 2.9.8          | 2.9.8           | 2.9.8           |
| p-bosh                    |                | 2.9.6-build.148 | 2.9.6-build.148 |
| p-healthwatch             | 1.8.1-build.1  | 1.8.1-build.1   | 1.8.1-build.1   |
|                           | 1.8.2-build.1  |                 |                 |
| pivotal-container-service | 1.7.0-build.26 | 1.7.0-build.26  | 1.7.0-build.26  |
| pivotal-telemetry-om      | 1.1.0-build.4  | 1.1.1-build.5   | 1.1.1-build.5   |
|                           | 1.1.1-build.5  |                 |                 |
+---------------------------+----------------+-----------------+-----------------+
```

File: presenters/table_presenter.go [hunk 2]
```diff
@@ -161,6 +161,79 @@ func (t TablePresenter) PresentPendingChanges(output api.PendingChangesOutput) {
...
+
+	for _, product := range products.ProductVersions {
+		moreAvailableProducts, firstProductLine := true, true
+		hasData := false
+
+		for moreAvailableProducts {
+			var outputData []string
+
+			moreAvailableProducts = false
+			if products.Available {
```
Comment (kcboyle) [inline presenters/table_presenter.go:189]: lost of nested `if`s in these `for` loops.
This is still clear, but aesthetically could put the logic of these if blocks into functions to make the nesting a little less crazy

**Comment by kcboyle** [comment]
Added a couple refactor suggestions in the commits @iplay88keys 
Nothing major, just a couple minor things if you're so inclined. 

_Thank you_ for re-generating docs 👍

File: presenters/table_presenter.go (hunk 2 above)
Comment (iplay88keys) [reply presenters/table_presenter.go:189]: Makes sense. I was definitely running short on time when doing this and wanted something working before the refactoring step.

File: commands/products.go (hunk 1 above)
Comment (iplay88keys) [reply commands/products.go:76]: The only reason I didn't pull these into separate functions was the complexity of their signatures when using Intellij's extract method feature. It might not have actually needed to be that complicated, though. I just didn't dig further at that point.

Comment (kcboyle) [reply commands/products.go:76]: had a quick additional commit to implement this.

**Comment by kcboyle** [comment]
Merged
